tty/pc-tty/ttypc_vga.c

```diff
@@ -1,119 +1,258 @@
 /* 
  * Phoenix-RTOS
  *
- * Operating system kernel
+ * VGA display
  *
- * ttypc VGA support (implemented after reading FreeBSD 4.4 pcvt driver)
- *
- * Copyright 2017, 2012, 2019 Phoenix Systems
  * Copyright 2008 Pawel Pisarczyk
+ * Copyright 2017, 2012, 2019, 2020 Phoenix Systems
  * Author: Pawel Pisarczyk, Lukasz Kosinski
  *
  * This file is part of Phoenix-RTOS.
  *
  * %LICENSE%
  */
 
-#include <stdio.h>
+#include <errno.h>
+#include <stdint.h>
 #include <string.h>
+
 #include <sys/io.h>
+#include <sys/minmax.h>
+#include <sys/mman.h>
 #include <sys/threads.h>
 
 #include "ttypc.h"
+#include "ttypc_vga.h"
+
+
+/* Reads from CRTC register */
+static uint16_t ttypc_vga_readreg(ttypc_t *ttypc, uint8_t reg)
+{
+	uint16_t val;
+
+	outb(ttypc->crtc, reg);
+	val = (uint16_t)inb(ttypc->crtc + 1) << 8;
+	outb(ttypc->crtc, reg + 1);
+	val |= inb(ttypc->crtc + 1);
+
+	return val;
+}
+
+
+/* Writes to CRTC register */
+static void ttypc_vga_writereg(ttypc_t *ttypc, uint8_t reg, uint16_t val)
+{
+	outb(ttypc->crtc, reg);
+	outb(ttypc->crtc + 1, (val >> 8) & 0xff);
+	outb(ttypc->crtc, reg + 1);
+	outb(ttypc->crtc + 1, (val >> 0) & 0xff);
+}
+
+
+/* Sets cursor type */
+static void _ttypc_vga_setcurt(ttypc_t *ttypc, uint8_t from, uint8_t to)
+{
+	uint16_t ctype = ttypc_vga_readreg(ttypc, CRTC_CURSTART);
+
+	ctype = (ctype & 0xc0e0) | ((uint16_t)from << 8 | to);
+	ttypc_vga_writereg(ttypc, CRTC_CURSTART, ctype);
+}
+
+
+void _ttypc_vga_switch(ttypc_vt_t *vt)
+{
+	ttypc_t *ttypc = vt->ttypc;
+	ttypc_vt_t *cvt = ttypc->vt;
+
+	if (vt == cvt)
+		return;
+
+	/* VGA memory -> VT memory */
+	memcpy(cvt->mem, cvt->vram, cvt->rows * cvt->cols * CHR_VGA);
```

Comment:
Side note - pointer to the screen buffer should be of pointer to volatile u16 type.

Reply:
All memory shared between application and hardware should be volatile, really.

Reply:
We use memcpy() and memsetw() for reading and writing to VGA memory. These functions discard volatile qualifier, so we would have to write custom functions to preserve it. Should I do it?

Reply:
General opinion is that memcpy is no suited for memory mapped devices:
https://stackoverflow.com/questions/36729240/passing-argument-2-of-memcpy-discards-volatile-qualifier-from-pointer-target

I think we need here our own memcpy implementation which operates on volatile buffer.